Start to fix touchpass

jgkamat opened the pull request:
Touchpass was broken from some point during last year in china, but I think I've fixed it (at least in the simulator for now). The cause of this was using capture in the receiving play, which I made some modificiations to fix. This is still a work in progress, but it's mostly done.

The main improvement I was thinking about was hijacking the 'fine-approach' state in capture to lunge towards the ball, as well as improve detecting when a touchpass fails earlier if possible. I've already modified course approach to suggest points to intercept on that allow for a touchpass.

Right now, only style is complaning, but I'll fix that once we have everything else wrapped up.

See #408

justbuchanan commented on soccer/gameplay/skills/touchball.py:
```diff
@@ -0,0 +1,49 @@
+import single_robot_behavior
```
This file should be named touch_ball.py to keep with the camel-case/underscore naming scheme

justbuchanan commented on soccer/gameplay/skills/touchball.py:
```diff
@@ -0,0 +1,49 @@
+        super().__init__(onlyApproach=True)
+
+    # Move back so we hit the mouth, not the side
+    adjDist = constants.Robot.Radius * 2
```
Constants typically start with a capital letter

justbuchanan commented on soccer/gameplay/skills/touchball.py:
```diff
@@ -0,0 +1,49 @@
+    # Move back so we hit the mouth, not the side
+    adjDist = constants.Robot.Radius * 2
+
+    ## Override so this so we only transition when the ball is in front
```
grammar/wording

justbuchanan commented on soccer/gameplay/skills/capture.py:
```diff
@@ -23,7 +23,10 @@ class State(Enum):
+    ## Capture Constructor
+    # onlyApproach - If true, any turning functions are turned off,
+    # useful for using capture to reflect/bounce moving ballls.
+    def __init__(self, onlyApproach=False):
```
Maybe rename `onlyApproach` to `faceBall` or something similar for clarity.  The name right now doesn't really convey that it affects facing and nothing else.

justbuchanan commented on soccer/gameplay/skills/touchball.py:
```diff
@@ -0,0 +1,49 @@
+    def find_intercept_point(self):
+        approach_vec = self.approach_vector()
+
+        adjFactor = robocup.Point(
```
This could also be written as `adjFactor = robocup.Point.direction(self.robot.angle) * -TouchBall.adjDist`

justbuchanan commented on soccer/gameplay/skills/angle_receive.py:
```diff
@@ -135,6 +136,17 @@ def execute_running(self):
 
+    # This depends on the superclasse's implementation somewhat
+    # If we stop using capture in pass_receive, this needs to be modified.
+    def on_enter_receiving(self):
```
It may be a little cleaner if we have PassReceive's constructor take in a capture behavior.  That way we don't have to override this transition and replace the behavior after it's been added.

jgkamat replied:
Pass Receive needs a fresh object every time it runs (this would break multiple coordinated passes, as in testCoordinatedPass). I could pass in a lambda that generates a Capture object. That probably would be a lot cleaner.

jgkamat replied:
actually this is a good idea, I'm going to do it.

justbuchanan commented:
Nice, it looks like this works a good bit better than before!

Looking through this, I wonder if it makes sense to move some of the capture overrides/changes into AngleReceive itself.  This would completely override PassReceive's implementation of the "receiving" state.  It seems like the main objective of the receiver after the ball is kicked is to face the appropriate angle and to stay on line with the ball's predicted trajectory.  This is pretty different from the capturing that PassReceive implements. Thoughts?

jgkamat commented:
Hmm, I think that would be a good idea, but I still would like to keep the touchball logic in its own play (Maybe skill?) that doesn't extend from capture. There are a lot of underlying similarities between this and capture, but the result is very different...

justbuchanan commented:
Sounds good, whatever works best.  Which parts of capture do you want to keep for TouchBall?

jgkamat commented:
Probably very little (I'm planning on having touchball only go to the nearest intersection point and stay there), so I'll probably not need to copy anything from capture. I'll see what happens though.

jgkamat commented:
I think I'm ready for round 2, I refactored all the touch-capture logic away, and I removed that mod I made to angle_receive that depended on its superclass.

I think some of the logic in touch_ball could be improved, but I think we should wait until we have real robots to test on before we assume anything additional about this behavior.

justbuchanan commented on soccer/gameplay/skills/capture.py:
```diff
@@ -101,11 +105,16 @@ def execute_course_approach(self):
         pos = self.find_intercept_point()
-        self.robot.face(main.ball().pos)
+
+        if (self.faceBall):
```
The facing logic may be better to put in `execute_running()` since it applies to all substates of `running`.

jgkamat replied:
yup, good idea

justbuchanan commented on soccer/gameplay/skills/touch_ball.py:
```diff
@@ -0,0 +1,140 @@
+            self.lastApproachTarget = pos
+
+    def on_exit_course_approach(self):
+        self.lastApproachTarget == None
```
Should this be `=` instead of `==`?

jgkamat replied:
aaah that should be a warning in python

justbuchanan commented:
Nice, looks good to me :+1:

jgkamat commented:
I think I've fixed the above issues, and I'm good on them. If you're good, you can merge it!
